src/nvim/buffer.c

```diff
@@ -3824,7 +3852,143 @@ int build_stl_str_hl(
   // add characters at the middle marker (if there is one) to
   // fill up the available space.
   } else if (width < maxwidth
-               && STRLEN(out) + maxwidth - width + 1 < outlen) {
+             && STRLEN(out) + maxwidth - width + 1 < outlen) {
+    // Find out if we are supposed to center the results,
+    //  or right / left align
+    // TODO(tjdevries): Place this somewhere else? More elegant?
+    int center_begin = -1;
+    int center_end = -1;
+    int middle_mark = -1;
+    bool center = false;
+
+    // Find at what index each of the centering actions takes place
+    for (int item_idx = 0; item_idx < itemcnt; item_idx++) {
+      if (item[item_idx].type == CenterBegin && center_begin == -1) {
+        center_begin = item_idx;
+      }
+      if (item[item_idx].type == CenterEnd && center_end == -1) {
+        center_end = item_idx;
+      }
+      if (item[item_idx].type == Middle && middle_mark == -1) {
+        middle_mark = item_idx;
+      }
+    }
+
+    // Only counts as a valid center if both pieces are found.
+    if (center_begin >= 0 && center_end >= 0) {
+      center = true;
+    }
+
+    // If we have a center area, then we deal with it now
+    if (center) {
+      // This is where we handle if the middle mark is presetn as well
+      if (middle_mark > 0) {
+        // Handle a combination of middle mark and centering
+        //
+        //       a                 b            c         d
+        // text\ %[this\ text\ here%=\ and\ some%]finishing
+        // text      his text here and some     finishing
+        // quad1             quad2|quad3            quad4
+        //       --fill1--                --fill2--
```

Comment:
This is a handy diagram.  It looks like you intended to have some of the variables used below reflected in the diagram, but the names don't match up.  Is `fill1` supposed to be the same as `fill_left`?